File: Tree/098.Validate-Binary-Search-Tree/098.Validate-Binary-Search-Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(
  root: TreeNode | null,
  max: number = Infinity,
  min: number = -Infinity
): boolean {
  if (root === null) {
    return true;
  }
  if (root.val > min && root.val < max) {
    if (
      isValidBST(root.left, root.val, min) &&
      isValidBST(root.right, max, root.val)
    ) {
      return true;
    }
  }
  return false;
}

File: Tree/098.Validate-Binary-Search-Tree/098.Validate-Binary-Search-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(root, minValue, maxValue):
            if root == None:
                return True
            
            return root.val > minValue and root.val < maxValue and dfs(root.left, minValue, root.val) and dfs(root.right, root.val, maxValue)
        return dfs(root, float('-inf'), float('inf'))
